### PR TITLE
Output space exploration

### LINKED ISSUE #154: Output Space Exploration using novelty search like optimization
OpenMole is a java package for exploring the behavior of simulation models. It contains an interesting [novelty search inspired way of exploring the model output space](https://openmole.org/PSE.html). It might be worthwhile to add a similar algorithm to the workbench.

In short, the n-dimensional output space is discretized (like with ε-archiving). For each grid cell, the total number of experiments in that box is tracked. For each generation of the evolution, selection proportionally favors grid cells with few solutions. The selected individuals are combined using a crossover operator and possibly mutated to create new experiments.

It should be relatively straightforward to set up this type of algorithm drawing on platypus-opt. It will involve a new method on the evaluator next to `perform_experiments`, `optimize`, and `robust_optimize`. In addition, a new custom selector to be used within a platypus-opt algorithm is required.

### PR DESCRIPTION
**Pull request by quaquel**
adds output space exploration to the workbench. 

- [x] unit tests and documentation are in place.
- [x] I might still want to update the example notebook
- [x] unit tests adds pytest-mock as a dependency in ci, and thus also for #166 

Closes #154.

### CODE REVIEW

**Comment by EwoutH** [comment]
Looks awesome! I will review it from a code-quality and user perspective tomorrow.

**Comment by EwoutH** [comment]
I started on a review, in general, it looks quite good!

Reviewing the docs on the other hand is quite a challenge without them begin built. Could you check if it's easy to enable doc building (see #171) before I review them?

**Comment by quaquel** [comment]
I have enabled the building of docs for pull requests. I guess we need one more commit here to trigger it.

**Comment by EwoutH** [comment]
If you can find the time, [cleaning up you commits](https://dev.to/aman_singh/how-to-cleanup-commit-history-using-git-reset-46p3) would be a great way to also trigger the CI! :)

Otherwise I will do some ci-trigger-magic when I review it (probably tomorrow).

**Comment by EwoutH** [comment]
Updated the CI and pyproject.toml with pytest-mock.

Read the Docs now runs, and generates the docs for this PR. However, for some reasone the outputspace_exploration module doesn't appear in it: https://emaworkbench--170.org.readthedocs.build/en/170/api_index.html#exploratory-modeling-framework

**Comment by quaquel** [comment]
Ok, I'll have a look later today. Probably missing links in some index files.

**Comment by EwoutH** [comment]
Perfect, take your time. I started on the review of the Jupyter example notebook, I hope to finish it later today.

**Comment by quaquel** [comment]
the docs now are displaying properly. I guess this is ready for a last cleanup of the commit history before being merged, unless you find something in the notebook

**Comment by EwoutH** [comment]
I just reviewed the notebook, and it's structured quite well! The main thing I'm missing is a bit of explanation and interpretation, from which the viewer can gain a better understanding of this method. It would add a lot of value if this example would be readable on itself, as a primer to OSE.

Reviewing Jupyter notebooks in GitHub turned out to be quite cumbersome, so I searched a bit and found an app, ReviewNB, which made this easy. I left some copies on the notebok, you can view and reply here (you probably have to login with GitHub): https://app.reviewnb.com/EwoutH/EMAworkbench/blob/outputspace_review/ema_workbench%2Fexamples%2Foutputspace_exploration_lakeproblem.ipynb

**Comment by quaquel** [comment]
That review app looks nice and the suggestions will indeed make the notebook a lot better. I'll try to make the suggested changes over the coming days.

**Comment by quaquel** [comment]
I have updated the notebook broadly in line with your suggestions. Note that I don't intend for these notebooks to be a tutorial on MOEAs so I have kept the additional text short but with links to relevant literature in case you want to know more.

I also added a small check to ensure the number of items in `grid_spec` matches the number of objectives (analogous to the problem with epsilon values in e-NSGA2). This in turn necessitated a small update to the test code.

**Comment by quaquel** [comment]
I cleaned the commit history again, so if you agree with the notebook update, this is ready to be merged.

**Review by EwoutH — CHANGES_REQUESTED** [review]
Code quality and structure looks good, I didn't encounter strange things.

Docs are also coming along nicely, I added a few comments on things I found unclear or ambiguous.

If you take a look at those, could you let me merge afterwards? I will squash it into a nice tidy commit.

File: ema_workbench/em_framework/optimization.py [hunk 1]
```diff
@@ -266,7 +266,9 @@ def to_dataframe(optimizer, dvnames, outcome_names):
     """
 
     solutions = []
-    for solution in platypus.unique(platypus.nondominated(optimizer.result)):
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/optimization.py:269]: First, it is redundant. Optimizer.results already returns the non dominated solutions. Second, it breaks with output space exploration, where we do not want to include directional information.

File: ema_workbench/em_framework/evaluators.py [hunk 2]
```diff
@@ -716,6 +719,7 @@ def optimize(
                         nfe between convergence check
     logging_freq : int
                    number of generations between logging of progress
+    variator : platypus GAOperator instance, optional
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/evaluators.py:722]: Not sure whether I want to go there. By defaulting to None, we trigger the default values within platypus-opt. The default in platypus-opt is Simulated Binary Crossover with Polynomial Mutation. Any custom platypus-opt GAOperator instance is possible, but what that means is covered by platypus-opt.

File: ema_workbench/em_framework/outputspace_exploration.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+                population_size=population_size,
+                generator=generator,
+                variator=variator,
+                **kwargs,
+            )
+        )
+
+
+class AutoAdaptiveOutputSpaceExploration(AdaptiveTimeContinuation):
+    """A combination of autoadaptive operator selection with OutputSpaceExploration.
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:308]: problem and grid spec are necessary. The rest has meaningful defaults

File: ema_workbench/em_framework/outputspace_exploration.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+from Chérel et al, we have taken the idea of the HitBox. Basically, this is an epsilon archive where one
+keeps track of how many solutions have fallen into each grid cell. Next, tournament selection based on novelty is
+used as the selective pressure. Novelty is defined as 1/nr. of solutions in same grid cell. This is then
+combined with auto-adaptive population sizing as used in e-NSGAII. This replaces the use of adaptive Cauchy mutation
+as used by Chérel et al. There is also an more sophisticated algorithm that adds auto-adaptive operator selection as
+used in BORG.
+
+The algorithm can be used in combination with the optimization functionality of the workbench.
+Just pass an OutputSpaceExploration instance as algorithm to optimize.
+
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:13]: If I know how, then yes. This probably overlaps with adding an overview page with links to all examples

File: ema_workbench/em_framework/outputspace_exploration.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    DifferentialEvolution,
+    SPX,
+    UM,
+    PCX,
+    UNDX,
+    Multimethod,
+)
+
+
+__all__ = ["OutputSpaceExploration", "AutoAdaptiveOutputSpaceExploration"]
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:44]: yes and yes.

File: ema_workbench/em_framework/outputspace_exploration.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+
+
+class OutputSpaceExploration(AdaptiveTimeContinuation):
+    """Basic genetic algorithm for output space exploration using novelty
+    search.
+
+    Parameters
+    ----------
+    problem : a platypus Problem instance
+    grid_spec : list of tuples
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:257]: Technically, you could get away with just the epsilon values. For cross seed comparisons, you would need the a value for the origin of the grid structure, and the epsilon for the size of the grid. I had left it at three for now, but we can change this to a minimum (or origin) and epsilon value.

File: ema_workbench/em_framework/outputspace_exploration.py [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                with min, max, and epsilon for
+                each outcome of interest
+    population_size : int, optional
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:260]: 100

File: ema_workbench/em_framework/optimization.py [hunk 8]
```diff
@@ -776,50 +778,49 @@ def mutate_categorical(self, child, i, type):  # @ReservedAssignment
     _mutate = {Real: mutate_real, Integer: mutate_integer, Subset: mutate_categorical}
 
 
-class CombinedMutator(CombinedVariator):
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/optimization.py:779]: yes that can go.

File: ema_workbench/em_framework/outputspace_exploration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:257]: By the way, which examples do you mean with tuples of 2 values?

File: ema_workbench/em_framework/outputspace_exploration.py [hunk 9]
```diff
@@ -0,0 +1,409 @@
+""" Provides a genetic algorithm based on novelty search for output space exploration.
+
+The algorithm is inspired by `Chérel et al (2015) <https://doi.org/10.1371/journal.pone.0138212>`_. In short,
+from Chérel et al, we have taken the idea of the HitBox. Basically, this is an epsilon archive where one
+keeps track of how many solutions have fallen into each grid cell. Next, tournament selection based on novelty is
+used as the selective pressure. Novelty is defined as 1/nr. of solutions in same grid cell. This is then
+combined with auto-adaptive population sizing as used in e-NSGAII. This replaces the use of adaptive Cauchy mutation
```
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:7]: yes, and link is being added

File: ema_workbench/em_framework/evaluators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EwoutH) [reply ema_workbench/em_framework/evaluators.py:722]: In that case I think I don't understand the _optional_. What happens if you leave `variator` unspecified?

File: ema_workbench/em_framework/outputspace_exploration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (EwoutH) [reply ema_workbench/em_framework/outputspace_exploration.py:257]: A few places:

- https://github.com/quaquel/EMAworkbench/pull/170/files#diff-17ddc728ebc309d5ce247a5cb98dcfbb88047b15dfb9bcab0877770da2c3656cR120
- https://github.com/quaquel/EMAworkbench/pull/170/files#diff-14e09aadf560c254302311b901de243dbe999d7fca44a1b4321b0dc0505e05b3R133-R134
- https://github.com/quaquel/EMAworkbench/pull/170/files#diff-427cb03187f4e1f151fcb26edaf2d2752df9faef498122e6b11d10d18c3d78c7R151

File: ema_workbench/em_framework/evaluators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/evaluators.py:722]: Then it takes its default value, which is None, as with any other keyword argument.

File: ema_workbench/em_framework/outputspace_exploration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (EwoutH) [reply ema_workbench/em_framework/outputspace_exploration.py:257]: I read `0, 1, 0.1` as `0.1, 0.1`, whoops...

Maybe format as
```Python
grid_spec = [
    (0, 1, 0.1),
    (0, 1, 0.1),
]
```

File: ema_workbench/em_framework/evaluators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (EwoutH) [reply ema_workbench/em_framework/evaluators.py:722]: So should we mention that it uses the platypus-opt default, which is Simulated Binary Crossover with Polynomial Mutation?

File: ema_workbench/em_framework/outputspace_exploration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/outputspace_exploration.py:257]: yes (see my black comment in #178 )

File: ema_workbench/em_framework/evaluators.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (quaquel) [reply ema_workbench/em_framework/evaluators.py:722]: I added a bit to the documentation

**Comment by quaquel** [comment]
I think this is now ready to be merged

**Comment by EwoutH** [comment]
Awesome, I agree!

Can I merge #178? Then I will throw in some CI, rebase and check if #170 passes, and then merge it.

**Review by EwoutH — APPROVED** [review]

**Comment by quaquel** [comment]
go ahead

**Comment by EwoutH** [comment]
@quaquel It's in, congratulations on an amazing new feature!

**Pull request merged**